# MongoDB with Python - PyMongo

Here is the source code for learning MongoDB with PyMongo using Python3. You can access the source code in [Google Colab](https://colab.research.google.com/drive/1m87G_ah_SNbTyw3c65vy2lki339I1dzv) or in this github. 

Before accessing MongoDB with Python, make sure you have the connection and internet available.

## 1. Install and Import Library

In [0]:
# install dnsrequest to access MongoDB with python, for this case you can install pymongon[srv] to your python env

#! pip install pymongo[srv]

In [0]:
# import library

import pymongo
from pymongo import MongoClient
import pprint
from IPython.display import clear_output

In [0]:
#client access to our main server MongoDB

client = MongoClient('mongodb+srv://userstudent:admin1234@cluster0-nnbxe.gcp.mongodb.net/test?retryWrites=true&w=majority')

#client access to our secondary server MongoDB
client2 = MongoClient('mongodb+srv://user1234:user1234@cluster0-1dcm4.gcp.mongodb.net/test?retryWrites=true&w=majority')

## 2. Database and Collection Overview

### **Our Main Client**

Before digging deeper let's see how our client (main and secondary) perform and what kind of database and collection they're keeping. Let's see and explore our main MongoDB client.

> **What am I going to do with this main client?**


*   I will use this database collection (***movies***) as comparison with my new collection model as requested from the practice case.
*   I wouldn't save my new collection here, because of some memory limit and I couldn't put or add more data to this cluster. So, this cluster will be used mainly for comparison.



In [8]:
# show all db in cluster

print(client.list_database_names())

['sample_mflix', 'admin', 'local']


In [9]:
db = client['sample_mflix']

# show all collection in used db
#print(db.list_collection_names())
pprint.pprint(db.list_collection_names())

['clean_movies_dwilaras',
 'clean_movies_nurlailiis',
 'clean_movies_devita',
 'movies_initial',
 'clean_movies_frans',
 'clean_movies_tara',
 'clean_movies_tunggal',
 'clean_movies',
 'users',
 'clean_movies_nurrimah',
 'movies_testtopik',
 'movies_scratch',
 'clean_movies_topik',
 'movies',
 'clean_movies_afifah',
 'clean_movies_andreas',
 'clean_movies_indra_1']


In [0]:
col_initial = db.movies

### **Our Secondary Client**

Before digging deeper let's see how our client secondary perform and what kind of database and collection they're keeping. Let's see and explore our main MongoDB client.

> **What am I going to do with this secondary client?**


*   I will use this cluster to store my new collection that will be named after **"clean_movies_isra"**.
*   I will connect to this cluster and use main cluster **"movies"** collection for final comparison.

In [11]:
# show all db in cluster
client2.list_database_names()

['movies_list', 'sample_mflix', 'admin', 'local']

In [12]:
db2 = client2['sample_mflix']

#Listing all the available collections
db2.list_collection_names()

['clean_movies_devita',
 'clean_movies_adipta',
 'movies_initial',
 'clean_movies_hania',
 'clean_movies_firli',
 'clean_movies_tio',
 'clean_movies_isra',
 'clean_movies_febi']

In [0]:
#Take 'movies_initial' collection for next observation
collect_init = db2['movies_initial']

## 3. Exploring The Surface

How we see our data collection definitely will help us to broaden our vision on our collected data. We will see how **movies** collection and **initial_movies** differs. We may see some differentiate, we may see some null values, so that we will be aware of our what to do next. For this case let's see our collection fields and keys value

In [62]:
#Checking 1 document from "movies" collection
col_initial.find_one()

{'_id': ObjectId('573a1390f29313caabcd5a93'),
 'awards': {'nominations': 0, 'text': '1 win.', 'wins': 1},
 'cast': ['Howard C. Hickman', 'Enid Markey', 'Lola May', 'Kate Bruce'],
 'countries': ['USA'],
 'directors': ['Reginald Barker',
  'Thomas H. Ince',
  'Raymond B. West',
  'Walter Edwards',
  'David Hartford',
  'Jay Hunt',
  'J. Parker Read Jr.'],
 'fullplot': 'Allegorical film about peace. A king starts a war, many of the women are against it, people are pressed into service. A count has constructed a submarine and gets the order to sink an ocean liner, that is also carrying - supposedly - ammunition for the enemy. The count refuses to fire the torpedos, and sinks the submarine. He survives, but in a limbo between death and life where he meets Jesus, who takes him over to preach peace. Naturally the king arrests him and sentences him to death for treason, but then Jesus shows him the real face of war.',
 'genres': ['Drama'],
 'imdb': {'id': 6517, 'rating': 6.3, 'votes': 162},
 '

In [63]:
#Checking 1 document from "movies_initial" collection
collect_init.find_one()

{'_id': ObjectId('5db003a2d0a60ca4acd0bee3'),
 'awards': '',
 'cast': 'Carmencita',
 'country': 'USA',
 'director': 'William K.L. Dickson',
 'fullplot': 'Performing on what looks like a small wooden stage, wearing a dress with a hoop skirt and white high-heeled pumps, Carmencita does a dance with kicks and twirls, a smile always on her face.',
 'genre': 'Documentary, Short',
 'imdbID': 1,
 'imdbRating': 5.9,
 'imdbVotes': 1032,
 'language': '',
 'lastupdated': '2015-08-26 00:03:45.040000000',
 'metacritic': '',
 'plot': 'Performing on what looks like a small wooden stage, wearing a dress with a hoop skirt and white high-heeled pumps, Carmencita does a dance with kicks and twirls, a smile always on her face.',
 'poster': 'http://ia.media-imdb.com/images/M/MV5BMjAzNDEwMzk3OV5BMl5BanBnXkFtZTcwOTk4OTM5Ng@@._V1_SX300.jpg',
 'rating': 'NOT RATED',
 'released': '',
 'runtime': '1 min',
 'title': 'Carmencita',
 'type': 'movie',
 'writer': '',
 'year': 1894}

#### Getting the Insight

From this we know that **movies** collection and **movies_inital** have some different. Start with,

1. `num_mflix_comment`
2. `tomatoes`
3. `released`
4. `awards`

where, for some case we will use ***nonEsixtinField***, ***datetime_strip***, and even maybe we will need ***arrayElemAt***  and ***split*** in our pipeline aggregation. To make sure, let's see how our collection's **dict_keys** for each **movies** and **movies_initial**.

In [64]:
lane = col_initial.find_one()
print(lane.keys())
print(len(lane.keys()))

dict_keys(['_id', 'plot', 'genres', 'runtime', 'cast', 'num_mflix_comments', 'poster', 'title', 'fullplot', 'countries', 'released', 'directors', 'writers', 'awards', 'lastupdated', 'year', 'imdb', 'type', 'tomatoes'])
19


In [65]:
lane1 = collect_init.find_one()
print(lane1.keys())
print(len(lane1.keys()))

dict_keys(['_id', 'imdbID', 'title', 'year', 'rating', 'runtime', 'genre', 'released', 'director', 'writer', 'cast', 'metacritic', 'imdbRating', 'imdbVotes', 'poster', 'plot', 'fullplot', 'language', 'country', 'awards', 'lastupdated', 'type'])
22


**Insight**

We saw it from queries aboove that our data is not match and not equal each other even from their **dict_keys** value and its collection's field value. So, for the aggregation we will need to build some new (maybe empty) field to match with **"movies"** collection.

In [66]:
datalsit = lane.keys()
if "num_mflix_comments" in datalsit:
  print("Row found")
else:
  print("Row not found")

Row found


In [67]:
collist = db.list_collection_names()
if "movies" in collist:
  print("Collection found")
else:
  print("Collection not found")

Collection found


## 4. Matching the Data Before Aggregating

In [0]:
#Looping pertama untuk mendapatkan imdb_list yang belum array formated.
imdb = list(col_initial.find({}, {'_id':0, 'imdb':1}))
imdb_list = []
for id in imdb:
  imdb_list.append(list(id.values())[0])

In [69]:
imdb_list[:5]

[{'id': 6517, 'rating': 6.3, 'votes': 162},
 {'id': 8133, 'rating': 7.8, 'votes': 4680},
 {'id': 12512, 'rating': 6.8, 'votes': 489},
 {'id': 13025, 'rating': 7.8, 'votes': 3738},
 {'id': 15498, 'rating': 7.1, 'votes': 327}]

## 5. Reshaping New Collection - Validating

### Projecting and Aggregating New Collection

In [0]:
#client = MongoClient('mongodb+srv://userstudent:admin1234@cluster0-nnbxe.gcp.mongodb.net/test?retryWrites=true&w=majority')


#========================================Projection============================================#
pipeline = [
    {
        #Every keys that I will include in the projection for my new collection'
        '$project': {
            
            #For some keys, I will use nonExistinField function to distinguished the "NULL" values if appear in the keys.
            
            'title': {"$cond": [{"$eq": ["$title", ""]}, "$nonExistinField", "$title"]},
            'year': {"$cond": [{"$eq": ["$year", ""]}, "$nonExistinField", "$year"]},
            'plot': {"$cond": [{"$eq": ["$plot", ""]}, "$nonExistinField", "$plot"]},
            'directors': {'$split': ["$director", ", "]},
            'cast': {'$split': ["$cast", ", "]},
            'writers': {'$split': ["$writer", ", "]},
            'languages': {'$split': ["$language", ", "]},
            'genres': {'$split': ["$genre", ", "]},
            'countries': {'$split': ["$country", ", "]},
            'fullplot': "$fullplot",
            'released': { '$concat': [ "datetime.datetime(", 
                                      {'$arrayElemAt':[{'$split': [ "$released", "-" ]}, 0]},
                                      ", ",
                                      {'$arrayElemAt':[{'$split': [ "$released", "-" ]}, 1]},
                                      ", ",
                                      {'$arrayElemAt':[{'$split': [ "$released", "-" ]}, 2]},
                                      ", 0, 0)" ] },
            'runtime': {"$cond": [{'$eq': ["$runtime", ""]}, 0 ,{'$arrayElemAt':[{'$split': [ "$runtime", " min" ]}, 0]}]},
            'num_mflix_comments': {'$sum':1},
            'poster': 1,
            'metacritic':{"$cond": [{"$eq": ["$metacritic", ""]}, "$nonExistinField", "$metacritic"]},
            'rated': "$rating",
            'imdb': {
                'id': "$imdbID",
                'rating': "$imdbRating",
                'votes': "$imdbVotes"
                },
            'awards': {'nominations': {'$arrayElemAt':[{'$split':[{'$arrayElemAt':[{'$split': [ "$awards", " nomination" ]}, -2]},' ']},-1]},
                       'text': "$awards",
                       'wins': {'$arrayElemAt':[{'$split':[{'$arrayElemAt':[{'$split': [ "$awards", " win" ]}, -2]},' ']},-1]}
                      },
            'type': {"$cond": [{"$eq": ["$type", ""]}, "$nonExistinField", "$type"]},
            'lastupdated': "$lastupdated",
            'tomatoes':{
                'lastupdated':'$lastupdated',
                'viewer': {'meter':'', 'numReviews':'', 'rating':''},
                'dvd':'',
                'fresh':'',
                'rotten':'',
                'production':''
                
            }
        }
    },
    
    
    #this match function will be playing as sorting function based on the imdb data that we have matched before
    { "$match": {"imdb": {"$in" : [dict(sorted(imdb_list[x].items(), key=lambda x: x[0])) for x in range(len(imdb_list))]}}}
    
    #we will named our new collection as 'clean_movies_isra'
    ,{'$out': 'clean_movies_isra'}
]

In [43]:
#Create new collection
collect_init.aggregate(pipeline)

In [70]:
#Check apakah collection sudah terbuat atau belum
db2['clean_movies_isra'].find_one()

{'_id': ObjectId('5db003a2d0a60ca4acd0bee5'),
 'awards': {'nominations': None, 'text': '1 win.', 'wins': '1'},
 'cast': ['Charles Kayser', 'John Ott'],
 'countries': ['USA'],
 'directors': ['William K.L. Dickson'],
 'fullplot': 'A stationary camera looks at a large anvil with a blacksmith behind it and one on either side. The smith in the middle draws a heated metal rod from the fire, places it on the anvil, and all three begin a rhythmic hammering. After several blows, the metal goes back in the fire. One smith pulls out a bottle of beer, and they each take a swig. Then, out comes the glowing metal and the hammering resumes.',
 'genres': ['Short'],
 'imdb': {'id': 5, 'rating': 6.2, 'votes': 1189},
 'languages': [''],
 'lastupdated': '2015-08-26 00:03:50.133000000',
 'num_mflix_comments': 1,
 'plot': 'Three men hammer on an anvil and pass a bottle of beer around.',
 'poster': '',
 'rated': 'UNRATED',
 'released': 'datetime.datetime(1893, 05, 09, 0, 0)',
 'runtime': '1',
 'title': 'Blac

### Validating New Collection

In [0]:
col_new = db2['clean_movies_isra']

In [72]:
lane_new = col_new.find_one()
print(lane_new.keys())
print(len(lane_new.keys()))

dict_keys(['_id', 'poster', 'awards', 'title', 'year', 'plot', 'directors', 'cast', 'writers', 'languages', 'genres', 'countries', 'fullplot', 'released', 'runtime', 'num_mflix_comments', 'rated', 'imdb', 'type', 'lastupdated', 'tomatoes'])
21


In [73]:
#Compare keys count dict between "clean_movies_isra" and "movies" collection

lane1 = collect_init.find_one()
print(lane1.keys())
print(len(lane1.keys()))

dict_keys(['_id', 'imdbID', 'title', 'year', 'rating', 'runtime', 'genre', 'released', 'director', 'writer', 'cast', 'metacritic', 'imdbRating', 'imdbVotes', 'poster', 'plot', 'fullplot', 'language', 'country', 'awards', 'lastupdated', 'type'])
22


**Insight**

From result above, we know that our new collection (**clean_movies_isra**) have same dict keys with **movies** collection as our comparison. For some case, ***imdbID*** is not included in our new collection because it's a field that already defined in **imdb** object keys.

**1. Validating - Banyak Document pada "clean_movies_isra" dan "movie"**

In [74]:
#Check jumlah document pada collection 'clean_movies_adipta'
len(list(db2['clean_movies_isra'].find()))

23539

In [75]:
#Check jumlah docoment pada collection 'movies'
len(list(db['movies'].find()))

23541

In [77]:
print("Jumlah dokumen pada `clean_movies_isra` sebanyak", len(list(db2['clean_movies_isra'].find())))
print("Jumlah dokumen pada `movies` sebanyak", len(list(db['movies'].find())))

Jumlah dokumen pada `clean_movies_isra` sebanyak 23539
Jumlah dokumen pada `movies` sebanyak 23541


**2. Validating - Semua Document pada "clean_movies_isra" dan "movie" Sama**

Based on the result above, we know that there are 2 documents that missing in **"clean_movies_isra"** (where ***movies*** have 23541 documents. Shall we analyze why and what is that missing document?

Let's us see what's going on here

In [0]:
list_clean = list(db2['clean_movies_isra'].aggregate([{'$sortByCount':'$imdb'}]))
list_movies = list(db['movies'].aggregate([{'$sortByCount':'$imdb'}]))

In [82]:
#Looping untuk melihat imdb mana yang berbeda.
for _id in list_movies:
  if _id not in list_clean:
    print(_id)

{'_id': {'id': 5, 'rating': 6.2, 'votes': 1189}, 'count': 2}


**3. Validating - Semua Fields pada "clean_movies_isra" ada pada "movie"**

In [89]:
#Print all fields in "clean_movies_isra"
field_isra = [list(doc.keys()) for doc in col_new.find({})]
field_isra = set([j for i in field_isra for j in i])
field_isra

{'_id',
 'awards',
 'cast',
 'countries',
 'directors',
 'fullplot',
 'genres',
 'imdb',
 'languages',
 'lastupdated',
 'metacritic',
 'num_mflix_comments',
 'plot',
 'poster',
 'rated',
 'released',
 'runtime',
 'title',
 'tomatoes',
 'type',
 'writers',
 'year'}

In [90]:
#Print all fields in "clean_movies_isra"
field_isra = [list(doc.keys()) for doc in col_new.find({})]
field_isra = set([j for i in field_isra for j in i])
field_isra

{'_id',
 'awards',
 'cast',
 'countries',
 'directors',
 'fullplot',
 'genres',
 'imdb',
 'languages',
 'lastupdated',
 'metacritic',
 'num_mflix_comments',
 'plot',
 'poster',
 'rated',
 'released',
 'runtime',
 'title',
 'tomatoes',
 'type',
 'writers',
 'year'}

In [91]:
field_movies = [list(doc.keys()) for doc in collect_init.find({})]
field_movies = set([j for i in field_movies for j in i])
field_movies

{'_id',
 'awards',
 'cast',
 'country',
 'director',
 'fullplot',
 'genre',
 'imdbID',
 'imdbRating',
 'imdbVotes',
 'language',
 'lastupdated',
 'metacritic',
 'plot',
 'poster',
 'rating',
 'released',
 'runtime',
 'title',
 'type',
 'writer',
 'year'}

In [92]:
len(field_movies)

22

In [93]:
len(field_isra)

22

**4. Validating - Document Values Sama**

In [18]:
db['movies'].find()[10]

{'_id': ObjectId('573a1391f29313caabcd71e3'),
 'awards': {'nominations': 0, 'text': '1 win.', 'wins': 1},
 'cast': ['Lois Wilson', 'Milton Sills', 'Theodore Roberts', 'Helen Ferguson'],
 'countries': ['USA'],
 'directors': ['William C. de Mille'],
 'fullplot': 'Wlliam deMille produced and directed Miss Lulu Bett, a film of extraordinary conviction and insight. It was then often the custom for unmarried women to lodge with family; thus we discover Miss Lulu in a boring Midwestern town, an exploited household drudge for her sister and her overbearing brother-in-law. In the course of the story (based upon the Pulitzer Prize play and novel by Zona Gale), Lulu evolves from slavery into an attractive and self-assured woman, prepared to make her own life. Revealed through wonderful performances and clever use of props, the characters are extraordinarily solid and involving.',
 'genres': ['Comedy', 'Drama'],
 'imdb': {'id': 12465, 'rating': 7.2, 'votes': 204},
 'lastupdated': '2015-08-21 00:15

In [16]:
list(db2['clean_movies_isra'].find({'title': 'Miss Lulu Bett'}))

[{'_id': ObjectId('5db003a2d0a60ca4acd0bfe5'),
  'awards': {'nominations': None, 'text': '1 win.', 'wins': '1'},
  'cast': ['Lois Wilson',
   'Milton Sills',
   'Theodore Roberts',
   'Helen Ferguson'],
  'countries': ['USA'],
  'directors': ['William C. de Mille'],
  'fullplot': 'Wlliam deMille produced and directed Miss Lulu Bett, a film of extraordinary conviction and insight. It was then often the custom for unmarried women to lodge with family; thus we discover Miss Lulu in a boring Midwestern town, an exploited household drudge for her sister and her overbearing brother-in-law. In the course of the story (based upon the Pulitzer Prize play and novel by Zona Gale), Lulu evolves from slavery into an attractive and self-assured woman, prepared to make her own life. Revealed through wonderful performances and clever use of props, the characters are extraordinarily solid and involving.',
  'genres': ['Comedy', 'Drama'],
  'imdb': {'id': 12465, 'rating': 7.2, 'votes': 204},
  'language